### PR TITLE
Metrics collection task

### PR DESCRIPTION
**Pull request by sriumcp**
Partly addresses iter8-tools/iter8#684

### CODE REVIEW

**Comment by huang195** [comment]
@sriumcp Is there a design doc to look at besides the PR?

**Comment by huang195** [comment]
This looks very fortio specific, but we're calling it `metrics/collect`. Would it make sense to call it more specifically `metrics/fortio` so there could be other `metrics/*` added in later?

File: lib/metrics/collect.go [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+type CollectTask struct {
+	Library string        `json:"library" yaml:"library"`
+	Task    string        `json:"task" yaml:"task"`
+	With    CollectInputs `json:"with" yaml:"with"`
+}
+
+// MakeCollect constructs a CollectTask out of a collect task spec
+func MakeCollect(t *v2alpha2.TaskSpec) (base.Task, error) {
+	if t.Task != "metrics/collect" {
+		return nil, errors.New("library and task need to be 'metrics' and 'collect'")
```
Comment (kalantar) [inline lib/metrics/collect.go:68]: Use constants `CollectTaskName` and `LibraryName` in both lines.

File: lib/metrics/collect.go [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+		err = json.Unmarshal(jsonBytes, &ct)
+	}
+	return ct, err
+}
+
+// InitializeDefaults sets default values for time duration and QPS for Fortio run
+// Default values are set only if the field is non-empty
+func (t *CollectTask) InitializeDefaults() {
+	if t.With.Time == nil {
+		t.With.Time = utils.StringPointer("5s")
```
Comment (kalantar) [inline lib/metrics/collect.go:87]: Use a constant for defaults so that they are easier to find and change.

File: lib/metrics/collect.go [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+	}
+	for i := 0; i < len(t.With.Versions); i++ {
+		if t.With.Versions[i].QPS == nil {
+			t.With.Versions[i].QPS = utils.Float32Pointer(8)
```
Comment (kalantar) [inline lib/metrics/collect.go:91]: Use a constant for defaults so that they are easier to find and change.

**Review by kalantar — CHANGES_REQUESTED** [review]
some small issues

**Comment by kalantar** [comment]
> This looks very fortio specific, but we're calling it `metrics/collect`. Would it make sense to call it more specifically `metrics/fortio` so there could be other `metrics/*` added in later?

It is fortio specific. However, if our purpose is to provide some build in metrics, I don't think we have to reveal the implementation solution we chose to adopt. In principle, we could change it without changing the name.

If we later choose to add alternatives, we can rename it or alias it at the time.

**Comment by kalantar** [comment]
I would like to be able to run this manually to try it. I can build a suitable handler image and include in an experiment. However, I need some help to understand what the metrics would look like.

I see some references in the test yaml, but how do define the metrics objects?

**Comment by sriumcp** [comment]
I agree with @kalantar on Fortio specificity. 

There is nothing Fortio specific about the `metrics/collect` specification / task. This is Iter8's way of surfacing built-in metrics. In particular, end-user documentation for this will not refer to Fortio.

The implementation happens to use Fortio.

**Comment by sriumcp** [comment]
@kalantar Changes are needed in analytics to complete this feature. FYI. This is merely attacking the handler portion of it.

**Review by kalantar — APPROVED** [review]

**Review by huang195 — APPROVED** [review]

**Pull request merged**